commons-mongo/jvm/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoClient.scala

```diff
@@ -0,0 +1,108 @@
+package com.avsystem.commons
+package mongo.typed
+
+import com.avsystem.commons.mongo.BsonValueInput
+import com.avsystem.commons.serialization.GenCodec
+import com.mongodb._
+import com.mongodb.connection.ClusterDescription
+import com.mongodb.reactivestreams.client.{MongoClient, MongoClients}
+import monix.eval.Task
+import monix.reactive.Observable
+import org.bson.Document
+
+import java.io.Closeable
+
+object TypedMongoClient {
+  def apply(): TypedMongoClient =
+    new TypedMongoClient(MongoClients.create())
+
+  def apply(connectionString: String): TypedMongoClient =
+    new TypedMongoClient(MongoClients.create(connectionString))
+
+  def apply(connectionString: ConnectionString): TypedMongoClient =
+    new TypedMongoClient(MongoClients.create(connectionString))
+
+  def apply(settings: MongoClientSettings): TypedMongoClient =
+    new TypedMongoClient(MongoClients.create(settings))
+
+  def apply(
+    connectionString: ConnectionString,
+    driverInformation: MongoDriverInformation,
+  ): TypedMongoClient =
+    new TypedMongoClient(MongoClients.create(connectionString, driverInformation))
+
+  def apply(
+    settings: MongoClientSettings,
+    driverInformation: MongoDriverInformation,
+  ): TypedMongoClient =
+    new TypedMongoClient(MongoClients.create(settings, driverInformation))
+}
+
+/**
+  * A better-typed wrapper over [[MongoClient]]. Uses Monix [[Task]] and [[Observable]] instead of
+  * [[org.reactivestreams.Publisher]]. Returns similar better-typed wrappers for database and client session objects.
+  */
+class TypedMongoClient(
```

Comment:
Observations:
1. in the beginning I was a little bit confused with the layers in this code. 
This 2 methods: withSession and inTransaction are IMO from different "layouts" (calling inTransaction, which creates new session in class, which can have other session is a little bit weird).
However it is consistent with the native mongo api, so it is ok for me. 
2. I am wondering if there is some option, to automate adding withSession, when calling inTransaction. Maybe something like this:
```
  def inTransaction[T](
    sessionOptions: ClientSessionOptions = ClientSessionOptions.builder().build(),
    transactionOptions: TransactionOptions = TransactionOptions.builder().build(),
  )(
    task: this.type => Task[T]
  ): Task[T] = ???
  ```
  however I am not convinced about this - I am just sharing thoughts.

Reply:
I already wrote `TypedMongoClient.inTransaction`, is that what you mean?

Reply:
No, TypedMongoClient.inTransaction requires to additionally write withSession. I am thinking how to serve less-general code, but with less requirements.